### DISCUSSION TITLE
Enhancing Data Structure Visibility in Laravel Collections with Proposed `xxxToArray()` Methods

### DISCUSSION THREAD

**Discussion by kayw-geek** [laravel/framework#50208] [Ideas]
Currently, some methods within the Collection class have their return types defined with generics in PHPDoc, which allows static analysis tools to recognize the specific data structures within the returned Collection. However, once the `toArray()` method is called, these data structures are transformed into `array<Tkey, mixed>`, resulting in the loss of the originally recognizable structure information. My idea is to add some `xxxToArray()` methods and accurately annotate them in PHPDoc, to avoid the need for manually adding inline PHPDoc in the code to indicate the array structure.

Example:
```PHP
 /**
     * Run an associative map over each of the items.
     *
     * The callback should return an associative array with a single key / value pair.
     *
     * @template TMapWithKeysKey of array-key
     * @template TMapWithKeysValue
     *
     * @param  callable(TModel, TKey): array<TMapWithKeysKey, TMapWithKeysValue>  $callback
     * @return \Illuminate\Support\Collection<TMapWithKeysKey, TMapWithKeysValue>|static<TMapWithKeysKey, TMapWithKeysValue>
     * @return array<TMapWithKeysKey, TMapWithKeysValue>
     */
    public function mapWithKeysToArray(callable $callback)
    {
        $this->mapWithKeys($callback)->toArray();
    }
```

**Comment by henzeb** [comment]
I don't think Taylor will be a big fan of those as it adds maintenance and the requirement of making a toArray method for each new method that returns a new collection instance with the modified data. Also, there are more places where type hinting gets lost in the Collection method chaining.

Looking at the code, the collect method on the Enumerable interface and EnumeratesValues class both use <TKey, TValue>, can't we just do the same with toArray on at least the Collection class?
